sys/net/include/netdev.h

```diff
@@ -0,0 +1,262 @@
+    NETDEV_ACTION_IDLE,
+    NETDEV_ACTION_RX,
+    NETDEV_ACTION_TX,               /**< only needed if pre-loading is enabled */
+} netdev_action_t;
+
+/**
+ * @brief   Definition of basic network device.
+ * @note    Feel free to expand if your device needs/supports more
+ */
+typedef enum {
+    NETDEV_STATE_OFF = 0,           /**< Device is powered off */
+    NETDEV_STATE_SLEEP,             /**< Device is sleeping */
+    NETDEV_STATE_IDLE,              /**< Device is idle */
+    NETDEV_STATE_RX,                /**< Device is in receive mode */
+    NETDEV_STATE_TX,                /**< Device is transmit mode */
+    NETDEV_STATE_PROMISCUOUS,       /**< Device is in receive mode and
+                                         accepts all packets without regard
+                                         for their destination */
+    NETDEV_STATE_TX_BURST,          /**< Device is burst sending and
+                                         does not accept packets */
+} netdev_state_t;
+
+/**
+ * @brief   The netdev data-structure holds the minimum information needed for
+ *          interaction with MAC layers
+ *
+ * The netdev structure is the parent for all network device driver descriptors.
+ */
+typedef struct netdev_t netdev_t;
+
+/**
+ * @brief   Receive data callback for data frames from given network device.
+ *
+ * @param[in] data          pointer to the received data in the packet buffer
+ */
+typedef void (*netdev_rcv_data_cb_t)(pkt_t *data);
+
+/**
+ * @brief   Network device API definition.
+ *
+ * @details This is a set of functions that must be implemented by any driver
+ *           for a network device.
+ */
+typedef struct {
+    /**
+     * @brief Send data via a given network device
+     *
+     * @param[in] dev               the network device
+     * @param[in] dest              the (hardware) destination address for the data
+     *                              in host byte order.
+     * @param[in] dest_len          the length of *dest* in byte
+     * @param[in] data              pointer to the data to sent
+     *
+     * @return  the number of byte actually send on success
+     * @return  -EAFNOSUPPORT if address of length dest_len is not supported
+     *          by the device *dev*
+     * @return  -ENODEV if *dev* is not recognized
+     * @return  -EMSGSIZE if the total frame size is too long to fit in a frame
+     *          of the device *dev*
+     * @return  a fitting negative other errno on other failure
+     */
+    int (*send_data)(netdev_t *dev, pkt_t *data);
+
+    /**
+     * @brief   Registers a receive data callback to a given network device.
+     *
+     * @param[in] dev   the network device.
+     * @param[in] cb    the callback.
+     *
+     * @return  0, on success
+     * @return  -ENODEV, if *dev* is not recognized
+     * @return  -ENOBUFS, if maximum number of registrable callbacks is exceeded
+     */
+    int (*add_receive_data_callback)(netdev_t *dev, netdev_rcv_data_cb_t cb);
```

Comment:
@haukepetersen The network device triggers an interrupt when transmission finished which OpenWSN needs for its internal state machine and time stamping.
@authmillenon I think we could also consider using `trigger` to MAC layer instead of `rcv_data_cb` as long as the network driver provides means to access the raw data received.